test/test_torch.py

```diff
@@ -12698,6 +12711,133 @@ def test_function_unwrap_message(self):
 
+    @staticmethod
+    def _test_unary_op_overlap(self, sz, data, op):
+        # c is identical to a:
+        c = data[0:sz]
+        a = data[0:sz]
+        c_exp = torch.empty_like(c)
+        op(a, out=c_exp)
+        op(a, out=c)
+        self.assertEqual(c_exp, c)
+
+        # c and a are independent:
+        c = data[0:sz]
+        a = data[sz:2 * sz]
+        c_exp = torch.empty_like(c)
+        op(a, out=c_exp)
+        op(a, out=c)
+        self.assertEqual(c_exp, c)
+
+        # c partially overlaps with a:
+        c = data[0:sz]
+        a = data[1:sz + 1]
+        with self.assertRaisesRegex(RuntimeError, 'unsupported operation'):
+            op(a, out=c)
+
+    def test_unary_op_input_output_overlap(self):
+        sz = 3
+        doubles = torch.randn(2 * sz)
+        positives = torch.randint(1, 100, (2 * sz,)).double()
+        ints = torch.randint(-100, 100, (2 * sz,))
+
+        self._test_unary_op_overlap(self, sz, doubles, torch.abs)
+        self._test_unary_op_overlap(self, sz, doubles, torch.acos)
+        self._test_unary_op_overlap(self, sz, doubles, torch.asin)
+        self._test_unary_op_overlap(self, sz, doubles, torch.atan)
+        self._test_unary_op_overlap(self, sz, ints, torch.bitwise_not)
+        self._test_unary_op_overlap(self, sz, doubles, torch.ceil)
+        self._test_unary_op_overlap(self, sz, doubles, torch.cos)
+        self._test_unary_op_overlap(self, sz, doubles, torch.erf)
+        self._test_unary_op_overlap(self, sz, doubles, torch.erfc)
+        self._test_unary_op_overlap(self, sz, doubles, torch.exp)
+        self._test_unary_op_overlap(self, sz, doubles, torch.expm1)
+        self._test_unary_op_overlap(self, sz, doubles, torch.floor)
+        self._test_unary_op_overlap(self, sz, doubles, torch.frac)
+        self._test_unary_op_overlap(self, sz, positives, torch.log)
+        self._test_unary_op_overlap(self, sz, positives, torch.log10)
+        self._test_unary_op_overlap(self, sz, positives, torch.log1p)
+        self._test_unary_op_overlap(self, sz, positives, torch.log2)
+        self._test_unary_op_overlap(self, sz, doubles, torch.neg)
+        self._test_unary_op_overlap(self, sz, doubles, torch.reciprocal)
+        self._test_unary_op_overlap(self, sz, doubles, torch.round)
+        self._test_unary_op_overlap(self, sz, positives, torch.rsqrt)
+        self._test_unary_op_overlap(self, sz, doubles, torch.sin)
+        self._test_unary_op_overlap(self, sz, doubles, torch.sigmoid)
+        self._test_unary_op_overlap(self, sz, doubles, torch.sqrt)
+        self._test_unary_op_overlap(self, sz, doubles, torch.tan)
+        self._test_unary_op_overlap(self, sz, doubles, torch.tanh)
+        self._test_unary_op_overlap(self, sz, doubles, torch.trunc)
+
+    @staticmethod
+    def _test_binary_op_input_output_overlap(self, op, device='cpu'):
+        sz = 3
+        data = torch.randn(3 * sz, device=device)
+
+        # c is identical to a and b:
+        c = data[0:sz]
```

Comment:
We can make this testing code less repetitive by adding a helper function:

```
def _test(op, c, a, b):
    c_exp = op(a, b)
    self.assertEqual(op(a, b, out=c), c_exp, op.__name__())
```

Reply:
addressed